rclcpp_components/cmake/rclcpp_components_register_node.cmake

```
@@ -67,6 +69,10 @@ macro(rclcpp_components_register_node target)
     "${_RCLCPP_COMPONENTS_${resource_index}__NODES}${component};${_path}/$<TARGET_FILE_NAME:${target}>\n")
   list(APPEND _RCLCPP_COMPONENTS_PACKAGE_RESOURCE_INDICES ${resource_index})
 
+  if(ARGS_NO_UNDEFINED_SYMBOLS)
+    target_link_options("${target}" PRIVATE "-Wl,--no-undefined")
```

This seems like a good idea, but won't be supported on Windows.

Maybe we could check that the compiler is GCC, or better yet check that the flag is supported before applying it?

```cmake
check_cxx_compiler_flag("-Wl,--no-undefined" linker_supports_no_undefined)
if (linker_supports_no_undefined)
    target_link_options(...)
endif()
```

https://cmake.org/cmake/help/latest/module/CheckCXXCompilerFlag.html

---

or something like this?

```suggestion
    if(ARGS_NO_UNDEFINED_SYMBOLS)
      if(WIN32)
        target_link_options("${target}" PRIVATE "/FORCE:UNRESOLVED")
      else()
        target_link_options("${target}" PRIVATE "-Wl,--no-undefined")
      endif()
    endif()
```

---

> The /FORCE linker option tells the linker to create an executable image (EXE file or DLL) even if a symbol is referenced but not defined or is defined more than once.

https://learn.microsoft.com/en-us/cpp/build/reference/force-force-file-output?view=msvc-170

it looks like this option is the exact opposite of what the GCC option does? I don't have a windows setup available for testing this.

> Maybe we could check that the compiler is GCC, or better yet check that the flag is supported before applying it?

I added the appropriate checks!